File: apps/js-sdk/firecrawl/src/v2/methods/scrape.ts
import {
  type Document,
  type ScrapeBrowserDeleteResponse,
  type ScrapeExecuteRequest,
  type ScrapeExecuteResponse,
  type ScrapeOptions,
} from "../types";
import { HttpClient } from "../utils/httpClient";
import { ensureValidScrapeOptions } from "../utils/validation";
import {
  throwForBadResponse,
  normalizeAxiosError,
} from "../utils/errorHandler";

/** Auto-resume bounds: a resume only ever follows the server's explicit
 * "processing continues" signal, and stops after this many resumes or
 * this much total sleeping — whichever comes first. */
const RESUME_MAX_ATTEMPTS = 5;
const RESUME_MAX_TOTAL_WAIT_MS = 20 * 60 * 1000;
const RESUME_MIN_DELAY_MS = 5 * 1000;
const RESUME_MAX_DELAY_MS = 10 * 60 * 1000;

/**
 * When a request failed because the document is still processing
 * server-side (large PDFs outlive their request window by design), the
 * error carries `details.state === "processing_continues"` plus a retry
 * delay. Returns that delay in ms, or undefined for every other error.
 */
export function processingContinuesDelayMs(err: unknown): number | undefined {
  const response = (err as { isAxiosError?: boolean; response?: any })
    ?.isAxiosError
    ? (err as { response?: any }).response
    : undefined;
  if (!response || response.status !== 408) return undefined;
  const body = response.data as
    | { code?: string; details?: { state?: string; retryAfterSeconds?: number } }
    | undefined;
  if (
    body?.code !== "SCRAPE_TIMEOUT" ||
    body?.details?.state !== "processing_continues"
  ) {
    return undefined;
  }
  const headerSeconds = Number(response.headers?.["retry-after"]);
  const seconds =
    typeof body.details.retryAfterSeconds === "number" &&
    Number.isFinite(body.details.retryAfterSeconds)
      ? body.details.retryAfterSeconds
      : Number.isFinite(headerSeconds)
        ? headerSeconds
        : 60;
  return Math.min(RESUME_MAX_DELAY_MS, Math.max(RESUME_MIN_DELAY_MS, seconds * 1000));
}

/**
 * Options for a direct scrape call. `autoResume` lives here — NOT on the
 * shared ScrapeOptions — because that type is reused verbatim inside
 * batch/crawl/search/extract payloads, where an SDK-only key would leak
 * onto the wire and fail strict endpoints.
 */
export type ScrapeCallOptions = ScrapeOptions & {
  /**
   * SDK-only (never sent to the API). Large documents (big PDFs) that
   * outlive the request window keep processing server-side; the API's
   * timeout error then carries `details.state === "processing_continues"`
   * and a Retry-After. When this is not `false`, the SDK sleeps that
   * long and re-issues the same request — the retry attaches to the
   * in-flight job and returns the finished result, so a large document
   * behaves like one slow successful call. Bounded (at most 5 resumes /
   * 20 minutes total wait); set `false` to surface the timeout error
   * immediately instead.
   */
  autoResume?: boolean;
};

export async function scrape(
  http: HttpClient,
  url: string,
  options?: ScrapeCallOptions,
  deps: { sleepImpl?: (ms: number) => Promise<void> } = {},
): Promise<Document> {
  const sleep =
    deps.sleepImpl ??
    ((ms: number) => new Promise<void>(resolve => setTimeout(resolve, ms)));
  if (!url || !url.trim()) {
    throw new Error("URL cannot be empty");
  }
  if (options) ensureValidScrapeOptions(options);

  // autoResume is SDK behavior, never part of the wire payload.
  const { autoResume, ...requestOptions } = options ?? {};
  const payload: Record<string, unknown> = { url: url.trim() };
  Object.assign(payload, requestOptions);

  // Per-request timeout. With auto-resume disabled this is exactly the
  // pre-existing behavior (explicit timeout +5s, else the client's own
  // default). With auto-resume on, the client-side timeout must outlast
  // the SERVER's wall — options.timeout, or the API's 5-minute default —
  // otherwise axios can abort in a photo-finish with the server's
  // processing_continues response and the resume signal is lost exactly
  // when it matters. The caller's configured client timeout is
  // respected as a floor, never overridden downward.
  const requestTimeout: { timeoutMs?: number } =
    autoResume === false
      ? typeof options?.timeout === "number"
        ? { timeoutMs: options.timeout + 5000 }
        : {}
      : {
          timeoutMs:
            Math.max(
              typeof options?.timeout === "number" ? options.timeout : 300_000,
              typeof http.getTimeoutMs === "function" ? http.getTimeoutMs() : 0,
            ) + 30_000,
        };

  let resumes = 0;
  let resumeWaitedMs = 0;
  while (true) {
    try {
      const res = await http.post<{
        success: boolean;
        data?: Document;
        error?: string;
      }>(
        "/v2/scrape",
        payload,
        requestTimeout,
      );
      if (res.status !== 200 || !res.data?.success) {
        throwForBadResponse(res, "scrape");
      }
      return (res.data.data || {}) as Document;
    } catch (err: any) {
      const delayMs = autoResume !== false ? processingContinuesDelayMs(err) : undefined;
      if (
        delayMs !== undefined &&
        resumes < RESUME_MAX_ATTEMPTS &&
        resumeWaitedMs + delayMs <= RESUME_MAX_TOTAL_WAIT_MS
      ) {
        // The document keeps processing server-side; the retry attaches
        // to the same in-flight job (content adoption) and returns the
        // finished result instead of restarting the work.
        resumes += 1;
        resumeWaitedMs += delayMs;
        await sleep(delayMs);
        continue;
      }
      if (err?.isAxiosError) return normalizeAxiosError(err, "scrape");
      throw err;
    }
  }
}

export async function interact(
  http: HttpClient,
  jobId: string,
  args: ScrapeExecuteRequest,
): Promise<ScrapeExecuteResponse> {
  if (!jobId || !jobId.trim()) {
    throw new Error("Job ID cannot be empty");
  }
  const hasCode = args?.code && args.code.trim();
  const hasPrompt = args?.prompt && args.prompt.trim();
  if (!hasCode && !hasPrompt) {
    throw new Error("Either 'code' or 'prompt' must be provided");
  }

  const body: Record<string, unknown> = {};
  if (hasCode) body.code = args.code;
  if (hasPrompt) body.prompt = args.prompt;
  body.language = args.language ?? "node";
  if (args.timeout != null) body.timeout = args.timeout;
  if (args.origin) body.origin = args.origin;

  try {
    const res = await http.post<ScrapeExecuteResponse>(
      `/v2/scrape/${jobId}/interact`,
      body,
      args.timeout != null ? { timeoutMs: args.timeout * 1000 + 5000 } : {},
    );
    if (res.status !== 200)
      throwForBadResponse(res, "interact with scrape browser");
    return res.data;
  } catch (err: any) {
    if (err?.isAxiosError)
      return normalizeAxiosError(err, "interact with scrape browser");
    throw err;
  }
}

export async function stopInteraction(
  http: HttpClient,
  jobId: string,
): Promise<ScrapeBrowserDeleteResponse> {
  if (!jobId || !jobId.trim()) {
    throw new Error("Job ID cannot be empty");
  }

  try {
    const res = await http.delete<ScrapeBrowserDeleteResponse>(
      `/v2/scrape/${jobId}/interact`,
    );
    if (res.status !== 200) throwForBadResponse(res, "stop interaction");
    return res.data;
  } catch (err: any) {
    if (err?.isAxiosError) return normalizeAxiosError(err, "stop interaction");
    throw err;
  }
}

/** @deprecated Use interact(). */
export async function scrapeExecute(
  http: HttpClient,
  jobId: string,
  args: ScrapeExecuteRequest,
): Promise<ScrapeExecuteResponse> {
  return interact(http, jobId, args);
}

/** @deprecated Use stopInteraction(). */
export async function stopInteractiveBrowser(
  http: HttpClient,
  jobId: string,
): Promise<ScrapeBrowserDeleteResponse> {
  return stopInteraction(http, jobId);
}

/** @deprecated Use stopInteraction(). */
export async function deleteScrapeBrowser(
  http: HttpClient,
  jobId: string,
): Promise<ScrapeBrowserDeleteResponse> {
  return stopInteraction(http, jobId);
}

File: apps/python-sdk/firecrawl/v2/methods/scrape.py
"""
Scraping functionality for Firecrawl v2 API.
"""

import time
import re
from uuid import uuid4
from typing import Optional, Dict, Any, List, Literal, Union
from ..types import (
    ScrapeOptions,
    Document,
    BrowserExecuteResponse,
    BrowserDeleteResponse,
    AlexandriaCall,
    AlexandriaScrapeData,
    AlexandriaScrapeResult,
)
from ..utils.normalize import normalize_document_input
from ..utils import FirecrawlError, HttpClient, handle_response_error, prepare_scrape_options, validate_scrape_options
from ..utils.auto_resume import ResumeTracker


def _prepare_scrape_request(url: str, options: Optional[ScrapeOptions] = None) -> Dict[str, Any]:
    """
    Prepare a scrape request payload for v2 API.
    
    Args:
        url: URL to scrape
        options: ScrapeOptions (snake_case) to convert and include
        
    Returns:
        Request payload dictionary with camelCase fields
    """
    if not url or not url.strip():
        raise ValueError("URL cannot be empty")

    request_data: Dict[str, Any] = {"url": url.strip()}

    if options is not None:
        validated = validate_scrape_options(options)
        if validated is not None:
            opts = prepare_scrape_options(validated)
            if opts:
                request_data.update(opts)

    return request_data



def scrape(
    client: HttpClient,
    url: str,
    options: Optional[ScrapeOptions] = None,
    *,
    auto_resume: Optional[bool] = None,
) -> Document:
    """
    Scrape a single URL and return the document.
    
    The v2 API returns: { success: boolean, data: Document }
    We surface just the Document to callers.
    
    Args:
        client: HTTP client instance
        url: URL to scrape
        options: Scraping options (snake_case)
        
    Returns:
        Document
    """
    payload = _prepare_scrape_request(url, options)

    resume = ResumeTracker(enabled=auto_resume is not False)
    while True:
        response = client.post("/v2/scrape", payload)

        if not response.ok:
            delay_s = resume.delay_or_none(response)
            if delay_s is not None:
                # The document keeps processing server-side; the retry
                # attaches to the same in-flight job (content adoption)
                # and returns the finished result.
                time.sleep(delay_s)
                continue
            handle_response_error(response, "scrape")

        body = response.json()
        if not body.get("success"):
            raise Exception(body.get("error", "Unknown error occurred"))

        document_data = body.get("data", {})
        normalized = normalize_document_input(document_data)
        return Document(**normalized)


MAX_ALEXANDRIA_CALLS = 10


def _prepare_scrape_alexandria_request(
    calls: List[Union[AlexandriaCall, Dict[str, Any]]],
    *,
    timeout: Optional[int] = None,
    integration: Optional[str] = None,
) -> Dict[str, Any]:
    if isinstance(calls, (dict, AlexandriaCall)):
        calls = [calls]
    if not calls:
        raise ValueError("At least one alexandria call is required")
    if len(calls) > MAX_ALEXANDRIA_CALLS:
        raise ValueError(f"At most {MAX_ALEXANDRIA_CALLS} alexandria calls are allowed per request")
    items: List[Dict[str, Any]] = []
    for call in calls:
        if isinstance(call, dict):
            call = AlexandriaCall(**call)
        elif not isinstance(call, AlexandriaCall):
            raise ValueError(f"Invalid alexandria call: {call!r}")
        provider = (call.provider or "").strip()
        capability = (call.capability or "").strip()
        if not provider:
            raise ValueError("Alexandria call provider cannot be empty")
        if not capability:
            raise ValueError("Alexandria call capability cannot be empty")
        item: Dict[str, Any] = {"provider": provider, "capability": capability}
        if call.options is not None:
            item["options"] = call.options
        items.append(item)
    payload: Dict[str, Any] = {"alexandria": items}
    if timeout is not None:
        if isinstance(timeout, bool) or not isinstance(timeout, int) or timeout <= 0:
            raise ValueError("Timeout must be a positive integer")
        payload["timeout"] = timeout
    if integration is not None and str(integration).strip():
        payload["integration"] = str(integration).strip()
    return payload


def _parse_scrape_alexandria_response(body: Dict[str, Any], request_id: str) -> AlexandriaScrapeData:
    data = body["data"]
    results = [AlexandriaScrapeResult(**item) for item in data["alexandria"]]
    return AlexandriaScrapeData(
        scrape_id=body.get("scrape_id"),
        alexandria=results,
        credits_cost=data["creditsCost"],
        request_id=request_id,
    )


def _alexandria_request_id(request_id: Optional[str]) -> str:
    value = str(uuid4()) if request_id is None else request_id
    if not re.fullmatch(r"[A-Za-z0-9._:-]{1,128}", value):
        raise ValueError("Invalid request_id")
    return value


def scrape_alexandria(client: HttpClient, calls, *, timeout: Optional[int] = None,
                    integration: Optional[str] = None, request_id: Optional[str] = None) -> AlexandriaScrapeData:
    payload = _prepare_scrape_alexandria_request(calls, timeout=timeout, integration=integration)
    request_id = _alexandria_request_id(request_id)
    headers = {**client._prepare_headers(), "x-request-id": request_id}
    try:
        response = client.post("/v2/scrape", payload, headers=headers,
                                    timeout=(min(timeout if timeout is not None else 50000, 50000) + 30000) / 1000)
        if response.status_code != 200 or not response.json().get("success"):
            handle_response_error(response, "scrape alexandria")
        return _parse_scrape_alexandria_response(response.json(), request_id)
    except FirecrawlError as error:
        error.request_id = request_id
        raise
    except Exception as error:
        raise FirecrawlError(str(error), request_id=request_id) from error


def interact(
    client: HttpClient,
    job_id: str,
    code: Optional[str] = None,
    *,
    prompt: Optional[str] = None,
    language: Literal["python", "node", "bash"] = "node",
    timeout: Optional[int] = None,
    origin: Optional[str] = None,
) -> BrowserExecuteResponse:
    """
    Interact with the scrape-bound browser session for a scrape job.

    Either ``code`` or ``prompt`` must be provided.  When ``prompt`` is given
    the server runs an AI agent that translates the natural-language instruction
    into browser actions.

    Args:
        client: HTTP client instance
        job_id: Scrape job ID
        code: Code to execute (optional if prompt is provided)
        prompt: Natural-language instruction for the browser agent (optional if code is provided)
        language: Programming language ("python", "node", or "bash")
        timeout: Execution timeout in seconds (1-300)
        origin: Optional request origin tag

    Returns:
        BrowserExecuteResponse with execution output
    """
    if not job_id or not job_id.strip():
        raise ValueError("Job ID cannot be empty")
    has_code = code and code.strip()
    has_prompt = prompt and prompt.strip()
    if not has_code and not has_prompt:
        raise ValueError("Either 'code' or 'prompt' must be provided")

    body: Dict[str, Any] = {
        "language": language,
    }
    if has_code:
        body["code"] = code
    if has_prompt:
        body["prompt"] = prompt
    if timeout is not None:
        body["timeout"] = timeout
    if origin is not None:
        body["origin"] = origin

    response = client.post(f"/v2/scrape/{job_id}/interact", body)
    if not response.ok:
        handle_response_error(response, "interact with scrape browser")

    payload = response.json()
    if not payload.get("success"):
        raise Exception(payload.get("error", "Unknown error occurred"))

    normalized = dict(payload)
    if "exitCode" in normalized and "exit_code" not in normalized:
        normalized["exit_code"] = normalized["exitCode"]
    if "cdpUrl" in normalized and "cdp_url" not in normalized:
        normalized["cdp_url"] = normalized["cdpUrl"]
    if "liveViewUrl" in normalized and "live_view_url" not in normalized:
        normalized["live_view_url"] = normalized["liveViewUrl"]
    if "interactiveLiveViewUrl" in normalized and "interactive_live_view_url" not in normalized:
        normalized["interactive_live_view_url"] = normalized["interactiveLiveViewUrl"]
    return BrowserExecuteResponse(**normalized)


def stop_interaction(
    client: HttpClient,
    job_id: str,
) -> BrowserDeleteResponse:
    """
    Stop the interaction session for a scrape job.

    Args:
        client: HTTP client instance
        job_id: Scrape job ID

    Returns:
        BrowserDeleteResponse
    """
    if not job_id or not job_id.strip():
        raise ValueError("Job ID cannot be empty")

    response = client.delete(f"/v2/scrape/{job_id}/interact")
    if not response.ok:
        handle_response_error(response, "stop interaction")

    payload = response.json()
    normalized = dict(payload)
    if "sessionDurationMs" in normalized and "session_duration_ms" not in normalized:
        normalized["session_duration_ms"] = normalized["sessionDurationMs"]
    if "creditsBilled" in normalized and "credits_billed" not in normalized:
        normalized["credits_billed"] = normalized["creditsBilled"]

    return BrowserDeleteResponse(**normalized)


def stop_interactive_browser(
    client: HttpClient,
    job_id: str,
) -> BrowserDeleteResponse:
    """Deprecated alias for stop_interaction()."""
    return stop_interaction(client, job_id)


def scrape_execute(
    client: HttpClient,
    job_id: str,
    code: Optional[str] = None,
    *,
    prompt: Optional[str] = None,
    language: Literal["python", "node", "bash"] = "node",
    timeout: Optional[int] = None,
    origin: Optional[str] = None,
) -> BrowserExecuteResponse:
    """Deprecated alias for interact()."""
    return interact(
        client,
        job_id,
        code,
        prompt=prompt,
        language=language,
        timeout=timeout,
        origin=origin,
    )


def delete_scrape_browser(
    client: HttpClient,
    job_id: str,
) -> BrowserDeleteResponse:
    """Deprecated alias for stop_interaction()."""
    return stop_interaction(client, job_id)
